internal/dig/shapes.lua

```diff
@@ -0,0 +1,289 @@
+-- shape definitions for gui/dig
+--@ module = true
+
+if not dfhack_flags.module then
+    qerror("this script cannot be called directly")
+end
+
+local to_pen = dfhack.pen.parse
+local CURSORS = {
+    INSIDE = { 1, 2 },
+    NORTH = { 1, 1 },
+    N_NUB = { 3, 2 },
+    S_NUB = { 4, 2 },
+    W_NUB = { 3, 1 },
+    E_NUB = { 5, 1 },
+    NE = { 2, 1 },
+    NW = { 0, 1 },
+    WEST = { 0, 2 },
+    EAST = { 2, 2 },
+    SW = { 0, 3 },
+    SOUTH = { 1, 3 },
+    SE = { 2, 3 },
+    VERT_NS = { 3, 3 },
+    VERT_EW = { 4, 1 },
+    POINT = { 4, 3 },
+}
+
+-- return the pen, alter based on if we want to display a corner and a mouse over corner
+function make_pen(direction, is_corner, is_mouse_over)
+    return to_pen {
+        ch = "X",
+        fg = COLOR_GREEN,
+        tile = dfhack.screen.findGraphicsTile(
+            "CURSORS",
+            direction[1],
+            direction[2] + (is_corner and 6 + (is_mouse_over and 3 or 0) or 0)
+        ),
+    }
+end
+
+-- Base shape class, should not be used directly
+Shape = defclass(Shape)
+Shape.ATTRS {
+    name = "",
+    arr = {},
+    has_point_fn = DEFAULT_NIL,
+    apply_options_fn = DEFAULT_NIL,
+    invert = false,
+}
+function Shape:update(width, height)
+    self.width = width
+    self.height = height
+    self.arr = {}
+    for x = 0, self.width do
+        self.arr[x] = {}
+        for y = 0, self.height do
+            local value = self:has_point_fn(x, y)
+            if not self.invert then
+                self.arr[x][y] = value
+            else
+                self.arr[x][y] = not value
+            end
+        end
+    end
+end
+
+-- Return a pen object based on the position of the x,y point within the shape
+-- Contains logic to determine which corner/side/etc... icon to draw to create an
+-- aesthetically pleasing border for the shape. Unfortunately the code itself
+-- very much not aesthetically pleasing but I'm not sure if that can be helped
```

Comment:
check out the Scrollbar tile calculations in library/lua/gui/widgets.lua sometime...